src/shogun/evaluation/Calibration.cpp

```diff
@@ -0,0 +1,244 @@
+
+		for (index_t j = 0; j < num_samples; ++j)
+		{
+			temp_confidences[j] += confidence_values[j];
+		}
+	}
+	for (index_t i = 0; i < num_classes; ++i)
+	{
+		SGVector<float64_t> confidence_values =
+		    result_labels->get_multiclass_confidences(i);
+		for (index_t j = 0; j < num_samples; ++j)
+		{
+			confidence_values[j] /= temp_confidences[j];
+		}
+		result_labels->set_multiclass_confidences(i, confidence_values);
+	}
+
+	return result_labels;
+}
+
+CBinaryLabels*
+CCalibration::apply_locked_binary(SGVector<index_t> subset_indices)
+{
+
+	CBinaryLabels* result_labels =
+	    (CBinaryLabels*)m_machine->apply_locked(subset_indices);
+	CCalibrationMethod* method =
+	    (CCalibrationMethod*)m_calibration_machines->get_element(0);
+	SGVector<float64_t> confidence_values =
+	    method->apply_binary(result_labels->get_values());
+	result_labels->set_values(confidence_values);
+
+	return result_labels;
+}
+
+EProblemType CCalibration::get_machine_problem_type() const
+{
+	return m_machine->get_machine_problem_type();
+}
+
+bool CCalibration::train(CFeatures* features)
+{
+	CCalibrationMethod* calibration_machine = NULL;
+	if (get_machine_problem_type() == PT_MULTICLASS)
+	{
+		SGVector<float64_t> confidences;
+		index_t num_calibration_machines =
+		    ((CMulticlassLabels*)get_labels())->get_num_classes();
+		m_calibration_machines =
+		    new CDynamicObjectArray(num_calibration_machines);
+		m_machine->train(features);
+		CMulticlassLabels* result_labels =
+		    (CMulticlassLabels*)m_machine->apply(features);
+		for (index_t i = 0; i < num_calibration_machines; ++i)
+		{
+			confidences = result_labels->get_multiclass_confidences(i);
+
+			calibration_machine = (CCalibrationMethod*)m_method->clone();
+			if (!calibration_machine->train(confidences))
+			{
+				return false;
+			}
+			m_calibration_machines->set_element(calibration_machine, i);
+		}
+	}
+	else
+	{
+		SGVector<float64_t> confidences;
+		m_calibration_machines = new CDynamicObjectArray(1);
+		m_machine->train(features);
+		CBinaryLabels* result_labels =
+		    (CBinaryLabels*)m_machine->apply_binary(features);
+
+		confidences = result_labels->get_values();
+
+		calibration_machine = (CCalibrationMethod*)m_method->clone();
+		if (!calibration_machine->train(confidences))
+		{
+			return false;
+		}
+		m_calibration_machines->set_element(calibration_machine, 0);
+	}
+
+	return true;
+}
+
+bool CCalibration::train_locked(SGVector<index_t> subset_indices)
+{
+	CBinaryLabels* m_result_labels =
+	    (CBinaryLabels*)m_machine->apply_locked(subset_indices);
+	CStatistics::SigmoidParamters params =
+	    CStatistics::fit_sigmoid(m_result_labels->get_values());
+	a = params.a;
+	b = params.b;
+
+	return true;
+}
+
+CCalibration::CCalibration() : CMachine()
+{
+	init();
+}
+
+void CCalibration::set_calibration_method(CCalibrationMethod* method)
+{
+	m_method = method;
+}
+
+void CCalibration::set_machine(CMachine* machine)
+{
+	m_machine = machine;
+}
+
+void CCalibration::init()
+{
```

Comment:
you are missing the `SG_ADD` macros for the class parameters. check other objects for examples